sigstore-java/src/main/java/dev/sigstore/http/UnsuccessfulResponseHandler.java

```diff
@@ -0,0 +1,118 @@
+
+/**
+ * Handler for 429 and standard server errors 5XX. Objects of this class are single use. A new
+ * instance should be created for each request (but this is handled in {@link
+ * HttpClients#newRequestFactory(HttpParams)}).
+ *
+ * <p>This implementation is a bit weird, but hopefully we can just move to grpc for rekor when
+ * ready.
+ */
+public class UnsuccessfulResponseHandler implements HttpUnsuccessfulResponseHandler {
+  private final Sleeper sleeper;
+
+  // for 5XX
+  private final BackOff exponentialBackOff;
+
+  // for 429
+  private static final int MAX_429_RETRIES = 10;
+  private int current429Retries = 0;
+
+  public static UnsuccessfulResponseHandler newUnsuccessfulResponseHandler() {
+    return new UnsuccessfulResponseHandler(Sleeper.DEFAULT, new ExponentialBackOff());
+  }
+
+  @VisibleForTesting
+  UnsuccessfulResponseHandler(Sleeper sleeper, ExponentialBackOff exponentialBackOff) {
+    this.sleeper = sleeper;
+    this.exponentialBackOff = exponentialBackOff;
+  }
+
+  @Override
+  public boolean handleResponse(HttpRequest request, HttpResponse response, boolean supportsRetry)
+      throws IOException {
+    if (!supportsRetry) {
+      return false;
+    }
+    try {
+      // exponential backoff for 5XX
+      if (response.getStatusCode() / 100 == 5) {
+        return BackOffUtils.next(sleeper, exponentialBackOff);
+      }
+      if (response.getStatusCode() == 429) {
+        return handle429Retry(response);
+      }
+      // we don't retry other codes
+    } catch (InterruptedException exception) {
+      // Mark thread as interrupted since we cannot throw InterruptedException here.
+      Thread.currentThread().interrupt();
+    }
+    return false;
+  }
+
+  private boolean handle429Retry(HttpResponse httpResponse)
+      throws InterruptedException, IOException {
+    if (current429Retries >= MAX_429_RETRIES) {
+      return false;
+    }
+    current429Retries++;
+
+    // Retry-After: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
+    var retryAfter = httpResponse.getHeaders().getRetryAfter();
+    if (retryAfter == null) {
+      // if there is no Retry-After just use exponentialBackoff (messes with counters, but that's
+      // ok)
+      return BackOffUtils.next(sleeper, exponentialBackOff);
+    }
+    var delay = calculateBackoff(retryAfter, Instant.now());
+    if (delay > 0) {
+      sleeper.sleep(delay);
+    }
+    return true;
+  }
+
+  // backoff in milliseconds
+  @VisibleForTesting
+  long calculateBackoff(String retryAfter, Instant now) throws IOException {
+    // parse as httpdate
+    var date = DateUtils.parseDate(retryAfter);
+    if (date != null) {
+      return date.toInstant().toEpochMilli() - now.toEpochMilli();
+    }
+
+    // maybe it is seconds
+    try {
+      return Long.parseLong(retryAfter) * 1000;
+    } catch (NumberFormatException nfe) {
+      // ignored
+    }
+    throw new IOException("Could not parse Retry-After");
```

Comment:
Should we just ignore unparseable `retry-after` instead of throwing an exception?

Reply:
I'm not sure exactly what to do there. I don't actually expect to get there unless rekor is misbehaving. Ignoring it would mean we wouldn't detect it? Maybe a warn log and delegate to exponential back off?

Reply:
If the number is unparseable, then just delegate to regular exponential back off as if `retry-after` was missing.

Reply:
Yup